hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/AtomicBatchHandler.java

```diff
@@ -85,18 +90,44 @@ public AtomicBatchHandler(
     /**
      * Performs checks independent of state or context.
      *
-     * @param context the context to check
+     * @param context the pure checks context
      */
     @Override
     public void pureChecks(@NonNull final PureChecksContext context) throws PreCheckException {
         requireNonNull(context);
-        final var op = context.body().atomicBatchOrThrow();
-        validateFalsePreCheck(context.body().hasBatchKey(), MISSING_BATCH_KEY);
-        for (final var transaction : op.transactions()) {
-            validateTruePreCheck(transaction.hasBody(), INVALID_TRANSACTION);
+        final TransactionBody txn = context.body();
+        requireNonNull(txn);
+        final AtomicBatchTransactionBody transactionBody = txn.atomicBatchOrThrow();
+
+        final List<Transaction> transactions = transactionBody.transactions();
+        requireNonNull(transactions);
+
+        if (transactions.isEmpty()) {
+            throw new PreCheckException(BATCH_LIST_EMPTY);
+        }
+
+        Set<TransactionID> set = new HashSet<>();
+        for (final var transaction : transactions) {
+            final TransactionBody txBody;
+            try {
+                txBody = bodyParser.apply(transaction);
+            } catch (HandleException e) {
+                throw new PreCheckException(e.getStatus());
+            }
+
+            validateTruePreCheck(set.add(txBody.transactionID()), BATCH_LIST_CONTAINS_DUPLICATES);
```

Comment:
Based on what Michael H said, we should not do these checks in pureChecks. It will fail in pre-handle the way its implemented currently

Reply:
We need to parse the inner transactions in order to do later validations of duplicate txs, missing batch keys, etc. In my understanding, what Michael H meant was that these kinds of checks should not happen before fees are charged and to fix that we need to move where pureChecks is called in the ingest workflow. These checks should still happen in pureChecks but where pureChecks is called should move, which is outside the scope of this PR.

Reply:
I mean checking `transactionID` is not required here. It will fail in `preHandle` for inner transactions @kimbor